### PR TITLE
[GLUTEN-8283][CH] Eliminate CSE via `ExpressionParser`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

Fixes: #8283

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

 unit tests

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 1]
```diff
@@ -255,16 +257,30 @@ std::pair<DB::DataTypePtr, DB::Field> LiteralParser::parse(const substrait::Expr
     return std::make_pair(std::move(type), std::move(field));
 }
 
-const DB::ActionsDAG::Node *
+const static std::string REUSE_COMMON_SUBEXPRESSION_CONF = "reuse_cse_in_expression_parser";
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:260]: blank line

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 2]
```diff
@@ -843,4 +868,88 @@ ExpressionParser::parseJsonTuple(const substrait::Expression_ScalarFunction & fu
...
+        return true;
+
+    if (a->type != b->type || !a->result_type->equals(*(b->result_type)) || a->children.size() != b->children.size()
+        || !a->isDeterministic() || !b->isDeterministic())
+        return false;
+
+    switch (a->type)
+    {
+        case DB::ActionsDAG::ActionType::INPUT: {
+            if (a->result_name != b->result_name)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:884]: should we compare result_type

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 3]
```diff
@@ -843,4 +868,88 @@ ExpressionParser::parseJsonTuple(const substrait::Expression_ScalarFunction & fu
...
+            return false;
+        }
+        case DB::ActionsDAG::ActionType::FUNCTION: {
+            if (!a->function_base->isDeterministic() || a->function_base->getName() != b->function_base->getName())
+                return false;
+
+            break;
+        }
+        default: {
+            LOG_DEBUG(getLogger("ExpressionParser"), "Unknow node type: {}|{}|{}", a->type, a->result_type->getName(), a->result_name);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:911]: "type:{}  result_type:{} name:{}"

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 4]
```diff
@@ -843,4 +868,88 @@ ExpressionParser::parseJsonTuple(const substrait::Expression_ScalarFunction & fu
...
+}
+
+// since each new node is added at the end of ActionsDAG::nodes, we expect to find the previous node and the new node will be dropped later.
+ExpressionParser::NodeRawConstPtr
+ExpressionParser::findFirstStructureEqualNode(NodeRawConstPtr node_, const DB::ActionsDAG & actions_dag) const
+{
+    for (const auto & node : actions_dag.getNodes())
+    {
+        if (node_ == &node)
+            continue;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:938]: blank line

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 5]
```diff
@@ -843,4 +868,88 @@ ExpressionParser::parseJsonTuple(const substrait::Expression_ScalarFunction & fu
...
+        a->result_name,
+        b->type,
+        b->result_type->getName(),
+        b->result_name);
+    return true;
+}
+
+// since each new node is added at the end of ActionsDAG::nodes, we expect to find the previous node and the new node will be dropped later.
+ExpressionParser::NodeRawConstPtr
+ExpressionParser::findFirstStructureEqualNode(NodeRawConstPtr node_, const DB::ActionsDAG & actions_dag) const
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:933]: maybe rename node_ to target is clearer

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 6]
```diff
@@ -843,4 +868,88 @@ ExpressionParser::parseJsonTuple(const substrait::Expression_ScalarFunction & fu
...
+{
+    for (const auto & node : actions_dag.getNodes())
+    {
+        if (node_ == &node)
+            continue;
+        if (areEqualNodes(node_, &node))
+        {
+            LOG_TRACE(
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:941]: LOG_TEST

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp (hunk 2 above)
Comment (lgbo-ustc) [reply cpp-ch/local-engine/Parser/ExpressionParser.cpp:884]: inputs have uniqu names

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp (hunk 3 above)
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:911]: LOG_WARNING

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp [hunk 7]
```diff
@@ -628,7 +643,18 @@ const DB::ActionsDAG::Node * ExpressionParser::toFunctionNode(
         std::string args_name = join(args, ',');
         result_name = ch_function_name + "(" + args_name + ")";
     }
-    return &actions_dag.addFunction(function_builder, args, result_name);
+    const auto * res_node = &actions_dag.addFunction(function_builder, args, result_name);
+    if (reuseCSE())
+    {
+        if (const auto * exists_node = findFirstStructureEqualNode(res_node, actions_dag))
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/ExpressionParser.cpp:649]: merge two ifs

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Parser/ExpressionParser.cpp (hunk 7 above)
Comment (lgbo-ustc) [reply cpp-ch/local-engine/Parser/ExpressionParser.cpp:649]:
```
        if (const auto * exists_node = findFirstStructureEqualNode(res_node, actions_dag))

```

**Review by baibaichen — APPROVED** [review]
LGTM

**Review by baibaichen — APPROVED** [review]
LGTM

**Review by taiyang-li — APPROVED** [review]

**Comment by lgbo-ustc** [comment]
A bench test as following

before
image

after
image

**Pull request merged**